File: 2_arrays_part2/2_merge_overlapping_subintervals.py
class Solution:
    def merge(self, nums: List[List[int]]) -> List[List[int]]:
        """
        1. Sort the list
        2. Create res list
        3. Iterate through list and compare the item with the last item of res to check if merging is possible
        4. If possible, update the last item of res, else add current item to res
        """
        nums.sort()
        res = [nums[0]]
        for i in range(1,len(nums)):
            if res[-1][-1] >= nums[i][0]:
                res[-1] = [res[-1][0], max(res[-1][1], nums[i][1])]
            else:
                res.append(nums[i])
        return res
        

File: 02_arrays_part2/2_merge_overlapping_subintervals.rs
use std::cmp::max;

impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;

        intervals.sort();
        let mut res: Vec<Vec<i32>> = vec![intervals[0].clone()];

        for interval in &intervals[1..] {
            let last = res.last().unwrap().to_vec();
            if interval[0] <= last[1] {
                res.pop();
                res.push(vec![last[0], max(last[1], interval[1])]);
            } else {
                res.push(interval.to_vec());
            }
        }
        res
    }
}
